src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -214,116 +154,91 @@ def visit_Subscript(self, node: foast.Subscript, **kwargs) -> foast.Subscript:
         )
 
     def _deduce_binop_type(
         self,
-        op: foast.BinaryOperator,
+        node: foast.BinOp,
         *,
-        parent: foast.BinOp,
-        left_type: ct.SymbolType,
-        right_type: ct.SymbolType,
+        left: foast.Expr,
+        right: foast.Expr,
         **kwargs,
     ) -> Optional[ct.SymbolType]:
-        if op in [
-            foast.BinaryOperator.ADD,
-            foast.BinaryOperator.SUB,
-            foast.BinaryOperator.MULT,
-            foast.BinaryOperator.DIV,
-        ]:
-            return self._deduce_arithmetic_binop_type(
-                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
-            )
-        else:
-            return self._deduce_logical_binop_type(
-                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
-            )
+        logical_ops = {foast.BinaryOperator.BIT_AND, foast.BinaryOperator.BIT_OR}
+        is_compatible = type_info.is_logical if node.op in logical_ops else type_info.is_arithmetic
 
-    def _deduce_arithmetic_binop_type(
-        self,
-        op: foast.BinaryOperator,
-        *,
-        parent: foast.BinOp,
-        left_type: ct.SymbolType,
-        right_type: ct.SymbolType,
-        **kwargs,
-    ) -> Optional[ct.SymbolType]:
-        left, right = TypeInfo(left_type), TypeInfo(right_type)
-
-        # if one type is `None` (not deduced, generic), we propagate `None`
-        if left.type is None or right.type is None:
-            return None
-
-        if (
-            left.is_arithmetic_compatible
-            and right.is_arithmetic_compatible
-            and are_broadcast_compatible(left, right)
-            and (broadcast_typeinfo := broadcast_typeinfos(left, right))
-        ):
-            return broadcast_typeinfo.type
+        # check both types compatible
+        for arg in (left, right):
+            if not is_compatible(arg.type):
+                raise FieldOperatorTypeDeductionError.from_foast_node(
+                    arg, msg=f"Type {arg.type} can not be used in operation '{node.op}'!"
+                )
+
+        if left.type == right.type:
+            return copy.copy(left.type)
+
+        self._check_operand_dtypes_match(node, left=left, right=right)
+
+        # check dimensions match and broadcast scalars to fields
+        for one_type, other_type in permutations([left.type, right.type]):
+            if type_info.can_promote_dims(other_type, one_type):
+                return copy.copy(one_type)
+
+        # the case of left_type == right_type is already handled above
+        # so here they must be incompatible
         raise FieldOperatorTypeDeductionError.from_foast_node(
-            parent,
-            msg=f"Incompatible type(s) for operator '{op}': {left.type}, {right.type}!",
+            node, msg=f"Incompatible dimensions in operation {left.type} '{node.op}' {right.type}!"
```

Comment:
When I started programming I was often confused by this style of error messages especially if other errors give me the actual syntax I wrote.

Reply:
Do you think the original `'{node.op}': {left.type}, {right.type}` is less confusing to a beginner?

Reply:
I would say so. Your latter sentence actually surprises me, if easy to do can you give an example? Here is what gcc does:
```
main.cpp:7:6: error: no match for ‘operator+’ (operand types are ‘std::array’ and ‘int’)
```
Which is more akin to the previous error message.

Reply:
And Python:
```
TypeError: unsupported operand type(s) for +: 'Bla' and 'int'
```

Reply:
That's good enough to change back I would say.